### PR TITLE
add expiry message when exam times out

### PR DESCRIPTION
**Pull request by ibrahimahmed443**
work in progress.

### CODE REVIEW

File: edx_proctoring/api.py [hunk 1]
```diff
@@ -1293,6 +1296,20 @@ def _get_timed_exam_view(exam, context, exam_id, user_id, course_id):
...
+        start_time = attempt['started_at']
+        end_time = attempt['completed_at']
+
+        if start_time and end_time:
+            attempt_duration_sec = (end_time - start_time).total_seconds()
+            allowed_duration_sec = attempt['allowed_time_limit_mins'] * 60
+            since_exam_ended_sec = (current_datetime - end_time).total_seconds()
+
+            # if the user took >= the available time, then the exam must have expired.
+            # but we show expiry message only when the exam was taken recently (less than SHOW_EXPIRY_MESSAGE_DURATION)
```
Comment (chrisndodge) [inline edx_proctoring/api.py]: Nice touch regarding SHOW_EXPIRY_MESSAGE_DURATION, I hadn't thought of that

**Comment by chrisndodge** [comment]
@muhhshoaib thanks. Also seems like we're missing some unit tests here to verify the use-case as well as confirming the additional messaging appears on the view HTML.

**Comment by chrisndodge** [comment]
@afzaledx @muhhshoaib can you do internal review?

File: edx_proctoring/tests/test_api.py [hunk 2]
```diff
@@ -1688,6 +1703,46 @@ def test_get_studentview_completed_timed_exam(self, status, expected_content):
...
+        rendered_response = get_student_view(
+            user_id=self.user_id,
+            course_id=self.course_id,
+            content_id=self.content_id_timed,
+            context={
+                'display_name': self.exam_name,
+            }
+        )
+
+        expected_content = 'The time allotted for this exam has expired.'
```
Comment (muhhshoaib) [inline edx_proctoring/tests/test_api.py]: Move this line in the setup method.

File: edx_proctoring/tests/test_api.py [hunk 3]
```diff
@@ -1688,6 +1703,46 @@ def test_get_studentview_completed_timed_exam(self, status, expected_content):
...
+        rendered_response = get_student_view(
+            user_id=self.user_id,
+            course_id=self.course_id,
+            content_id=self.content_id_timed,
+            context={
+                'display_name': self.exam_name,
+            }
+        )
+
+        expected_content = 'You have submitted your timed exam.'
```
Comment (muhhshoaib) [inline edx_proctoring/tests/test_api.py]: Move this line in the setup method with the suitable name.

**Comment by muhhshoaib** [comment]
once comments addressed. :+1: LGTM

**Comment by chrisndodge** [comment]
squashed commits

**Pull request merged**